frostsnap_coordinator/src/bitcoin/wallet.rs

```diff
@@ -185,6 +236,91 @@ impl CoordSuperWallet {
         Ok(self.address_info(master_appkey, &spk, index, keychain))
     }
 
+    pub fn next_unused_address(&mut self, master_appkey: MasterAppkey) -> Result<AddressInfo> {
+        self.lazily_initialize_key(master_appkey);
+        let reserved_spks = self.reserved_spks.as_ref();
+        let keychain = BitcoinAccountKeychain::external();
+        let (index, spk) = {
+            let mut db = self.db.lock().unwrap();
+            self.tx_graph.mutate(&mut *db, |tx_graph| {
+                let mut changeset = WalletIndexedTxGraphChangeSet::default();
+                loop {
+                    let ((i, spk), changes) = tx_graph
+                        .index
+                        .next_unused_spk((master_appkey, keychain))
+                        .ok_or(anyhow!(
+                            "exhausted derivation indices for {}'s external keychain",
+                            master_appkey
+                        ))?;
+                    changeset.merge(changes.into());
+                    if reserved_spks.contains(&spk) {
+                        tx_graph.index.mark_used((master_appkey, keychain), i);
+                        println!("contains spk!");
+                        continue;
+                    }
+                    return Ok(((i, spk), changeset));
+                }
+            })?
+        };
+        Ok(self.address_info(master_appkey, &spk, index, keychain))
+    }
+
+    pub fn mark_address_shared(
+        &mut self,
+        master_appkey: MasterAppkey,
+        derivation_index: u32,
+    ) -> Result<bool> {
+        self.lazily_initialize_key(master_appkey);
+        let keychain = BitcoinAccountKeychain::external();
+        let mut db = self.db.lock().unwrap();
+        self.tx_graph
+            .multi(&mut self.reserved_spks)
+            .mutate(&mut db, |tx_graph, reserved_spks| {
+                let mut tx_graph_changeset = WalletIndexedTxGraphChangeSet::default();
+                let mut reserved_spks_changeset = ReservedSpksChangeSet::default();
+                match tx_graph
+                    .index
+                    .reveal_to_target((master_appkey, keychain), derivation_index)
+                {
+                    Some((_, changeset)) => tx_graph_changeset.merge(changeset.into()),
+                    None => return Ok((false, (tx_graph_changeset, reserved_spks_changeset))),
```

Comment:
should reveal_to_target `None` map to error?

Reply:
I guess it can. Also at line `293`. However, I tend to avoid errors in general because we are not `try-catch`-ing them on the dart side. So if it ever errors, the UI state will not progress and the user will be confused.

Reply:
This particular scenario would involve something going seriously wrong, I think we'd rather freeze than silently continue